ir/memory.cpp

```diff
@@ -195,6 +195,125 @@ ostream& operator<<(ostream &os, const Pointer &p) {
 }
 
 
+Byte::Byte(const Memory &m, expr &&byterepr): m(m), p(move(byterepr)) {
+  assert(p.bits() == m.bitsByte());
+}
+
+Byte::Byte(const Pointer &ptr, unsigned i, expr non_poison)
+    : m(ptr.getMemory()) {
+  // TODO: support pointers larger than 64 bits.
+  assert(m.bitsMemorySize() <= 64 && m.bitsMemorySize() % 8 == 0);
+  // non_poison should be either bool or 1 bit.
+  if (non_poison.isBool())
+    non_poison = non_poison.toBVBool();
+  assert(non_poison.bits() == 1);
+
+  expr byteofs = expr::mkUInt(i, 3);
+  p = expr::mkUInt(1, 1).concat(non_poison).concat(ptr()).concat(byteofs);
+  assert(p.bits() == m.bitsByte());
+}
+
+Byte::Byte(const Memory &m, expr data, expr non_poison): m(m) {
+  assert(m.bitsByte() >= 1 + 8 + 8);
+  assert(data.bits() == 8);
+  assert(non_poison.bits() == 8);
+
+  unsigned padding = m.bitsByte() - 1 - 8 - 8;
+  p = expr::mkUInt(0, 1).concat(expr::mkUInt(0, padding)).concat(non_poison)
+                        .concat(data);
+}
+
+
+void valueToBytes(const StateValue &val, const Type &fromType,
+                  const Memory &mem, vector<Byte> &bytes) {
+  if (fromType.isPtrType()) {
+    Pointer p(mem, val.value);
+    unsigned bytesize = mem.bitsMemorySize() / 8;
+
+    for (unsigned i = 0; i < bytesize; ++i)
+      bytes.emplace_back(p, i, val.non_poison);
+  } else {
+    StateValue bvval = fromType.toBV(val);
+    unsigned bitsize = bvval.bits();
+    unsigned bytesize = divide_up(bitsize, 8);
+
+    bvval = bvval.zext(bytesize * 8 - bitsize);
+
+    for (unsigned i = 0; i < bytesize; ++i) {
+      expr data  = bvval.value.extract((i + 1) * 8 - 1, i * 8);
+      expr np    = bvval.non_poison.extract((i + 1) * 8 - 1, i * 8);
```

Comment:
StateValue has extract() now. You can do these 2 lines in one go.

Reply:
I couldn't find it - StateValue's extract is not pushed to master yet?

Reply:
ah, never mind. please ignore